[REEF-1469] Simplify logic in OperatorTopology

afchung opened the pull request:
This addressed the issue by
  * Replacing logic in GetNodesWithData with synchronous calls to GetData. This is okay because we will need to wait for all children nodes to receive data before proceeding.
  * Renaming variables for clarity.
  * Adding ChildNodeContainer to simplify management of children nodes.

JIRA:
  [REEF-1469](https://issues.apache.org/jira/browse/REEF-1469)

afchung commented:
@dkm2110 would be great if you could take a look. Thanks!

markusweimer commented on lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/ChildNodeContainer.cs:
```diff
@@ -0,0 +1,92 @@
+        /// </summary>
+        public IEnumerable<T> GetDataFromAllChildren()
+        {
+            return this.SelectMany(child => child.GetData());
```
Does the return of `IEnumerable` here break the chaining of LINQ queries? Shouldn't it be `IQueryable`?

afchung replied:
In this case it shouldn't make a difference: http://stackoverflow.com/questions/2876616/returning-ienumerablet-vs-iqueryablet

afchung replied:
The caller can also cast `IEnumerable` to `IQueryable` with `.AsQueryable` if needed.

jwang98052 commented on lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/OperatorTopology.cs:
```diff
@@ -51,13 +50,10 @@ public sealed class OperatorTopology<T> : IOperatorTopology<T>, IObserver<Genera
-        private readonly Dictionary<string, NodeStruct<T>> _idToNodeMap;
         private readonly INameClient _nameClient;
         private readonly Sender _sender;
-        private readonly BlockingCollection<NodeStruct<T>> _nodesWithData;
```
Do you see any potential impact for removing NodeWithData BlockingCollection?

afchung replied:
No

jwang98052 commented on lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/OperatorTopology.cs:
```diff
@@ -355,78 +318,13 @@ public void OnCompleted()
-                        return nodesSubsetWithData;
-                    }
-
-                    while (_nodesWithData.Count != 0)
```
Do you know why it want to drain the queue here?

afchung replied:
It tires to remove irrelevant nodes such as the parent node from the queue. In the new code, we are not paying attention to those nodes at all.

afchung replied:
@dkm2110 please let me know if I explained this incorrectly.

dkm2110 replied:
You did

dkm2110 commented on lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/OperatorTopology.cs:
```diff
@@ -322,27 +305,7 @@ public T ReceiveFromChildren(IReduceFunction<T> reduceFunction)
-            }
-
-            return reduceFunction.Reduce(receivedData);
+            return reduceFunction.Reduce(_childNodeContainer.GetDataFromAllChildren());
```
This can potentially significantly increase the memory footprint, right? Consider the case where we are not using pipelining and T is a large vector that we want to reduce. We now wait for all of them to be present before doing reduce. For flat topology with P nodes there will be P such large vectors, although one can argue that we should use binary topology then :) . Since we in any case are waiting for all children and reduce is assumed to be associative and commutative, I am wondering if we should apply reduce incrementally one child at a time. This can potentially lead to smaller memory footprints. Does that make sense?

afchung replied:
The logic here is the same as before. Applying reduce one child at a time also increases network overhead. I think this merits another JIRA item.

dkm2110 replied:
@afchung  can you explain in a little bit of detail why there will be increased network overhead? I am ok addressing it in separate JIRA. This will hit user only if he does not use group communication in an optimal way.

afchung replied:
@dkm2110 Sorry I have misunderstood it. This should not increase network overhead, but I think we should defer this to another JIRA item.

dkm2110 commented:
@afchung left one comment. Otherwise LGTM.

dkm2110 commented:
I again see that appveyor build is failing.

jwang98052 commented:
I completed a pass.

afchung commented:
@dkm2110 @jwang98052 @markusweimer thanks for the review! I've addressed your comments, please have another look. The Appveyor build failures are known failures and the code change here should not have affected the failed tests.

dkm2110 commented:
LGTM to merge

jwang98052 commented:
LGTM.
